core/src/main/java/io/kestra/core/http/client/apache/FailedResponseInterceptor.java

```
@@ -0,0 +1,33 @@
+package io.kestra.core.http.client.apache;
+
+import io.kestra.core.http.HttpResponse;
+import io.kestra.core.http.HttpService;
+import io.kestra.core.http.client.HttpClientResponseException;
+import lombok.AllArgsConstructor;
+import org.apache.hc.core5.http.EntityDetails;
+import org.apache.hc.core5.http.HttpEntityContainer;
+import org.apache.hc.core5.http.HttpException;
+import org.apache.hc.core5.http.HttpResponseInterceptor;
+import org.apache.hc.core5.http.protocol.HttpContext;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+
+@AllArgsConstructor
+public class FailedResponseInterceptor implements HttpResponseInterceptor {
+    @Override
+    public void process(org.apache.hc.core5.http.HttpResponse response, EntityDetails entity, HttpContext context) throws HttpException, IOException {
+        if (response.getCode() >= 400) {
+            String error = "Failed http request with response code '" + response.getCode() + "'";
```

I don't know if it has to be done like that, but I am used to have a client return null for 404 status instead of exception.

---

It's was the behavior of Micronaut client, and I found it better and safer, any response code >= 400 need a special behavior to be catch, exception is explicit, validating response code on each call can be bug prone

---

I'm also for returning null for 404.

As far as I remember, this was why we added a special boolean to allowEmptyResponse to return null, to bypass 404 throwing an exception.

---

I'm stronglmy against changing this behavior, it's a breaking change, and people need to manage the error, we couldn't let them failed silently.